core/src/nucleus/ribosome/run_dna.rs

```diff
@@ -0,0 +1,176 @@
+use context::Context;
+use holochain_core_types::error::{HcResult, HolochainError};
+use holochain_wasm_utils::{
+    error::{RibosomeErrorCode, RibosomeReturnCode},
+    memory_allocation::decode_encoded_allocation,
+};
+use nucleus::{
+    ribosome::{api::ZomeApiFunction, memory::SinglePageManager, Runtime},
+    ZomeFnCall, ZomeFnResult,
+};
+use std::{str::FromStr, sync::Arc};
+use wasmi::{
+    self, Error as InterpreterError, FuncInstance, FuncRef, ImportsBuilder, ModuleImportResolver,
+    ModuleInstance, NopExternals, RuntimeValue, Signature, ValueType,
+};
+
+/// Executes an exposed zome function in a wasm binary.
+/// Multithreaded function
+/// panics if wasm binary isn't valid.
+pub fn run_dna(
+    dna_name: &str,
+    context: Arc<Context>,
+    wasm: Vec<u8>,
+    zome_call: &ZomeFnCall,
+    parameters: Option<Vec<u8>>,
+) -> ZomeFnResult {
```

Comment:
This function is such a monster lol

Reply:
Gradually getting better though

Reply:
yep. There is no intent to make it smaller, thats why I'm giving it its own file.
I did shrunk it a bit my removing that Runtime impl.